treewide: Default to C17 as the minimum required C standard

carlescufi opened the pull request:
C99 has been the minimum required C standard version for Zephyr since its inception. After multiple attempts and discussions, a decision has been made to upgrade to C17 going forward.
This commits replaces the default C standard from C99 to C17 in the configuration and build system, and deprecates support for the older standards.


Closes #97061

carlescufi commented:
@stephanosio @nashif @keith-packard should we also enforce C++17 at the same time, for consistency?

nvetrini commented:
@nashif  You may want to consider switching to MISRA C:2012 Amendment 3 as a basis for the guidelines, since that edition of the guidelines have much better coverage of C17 than Amendment 2 (e.g. it has guidelines for the safe use of `_Generic`, which are not present in Amendment 2)

tejlmand commented on cmake/compiler/compiler_features.cmake:
good catch.

Yes they should remain, cause they go into a list to tell CMake what standard features are available.
And there we should include old ones.

Some (external) code might do: `target_compile_features(foo PRIVATE c_std_99)` which is perfectly valid cause that just means code requires feature from c99, which is also supported by c17 (our new default).

tejlmand reviewed (changes requested):
please keep code in compiler_features.cmake.

The list should still include old standards as that will allow downstream users to request those as required (and it's support when using newer c17 std)

kartben commented:
> a decision has been made to upgrade to C17 going forward.

@carlescufi for tracking purposes, can you please link to the decision in question? I am not questioning that a decision has been made, but I am having a hard time tracking down when/where so I can only imagine it's even harder for mere mortals to find that information.

stephanosio commented on doc/develop/languages/c/index.rst:
```diff
@@ -35,16 +35,14 @@ toolchain that supports the C99 standard and above:
+2017 version of the standard (ISO/IEC 9899:2017):
+
+* _Generic keyword
+* _Static_assert keyword
```
`_Generic` and `_Static_assert` were introduced in C11, not C17.

stephanosio commented:
> @stephanosio @nashif @keith-packard should we also enforce C++17 at the same time, for consistency?

That is probably not a good idea since C++ versions are not necessarily in sync with the C versions, and C++17 is not a mere bug fix over the previous C++11 standard; in fact, [C++17 even removes some existing features](https://en.wikipedia.org/wiki/C%2B%2B17#Removed_features).

As for dropping support for the archaic C++ standards prior to C++11 (i.e. C++98 and C++03), that would be reasonable since C++11 introduces many C99 features required by Zephyr.

carlescufi committed:
c standard: Default to C17 as the minimum required C standard

C99 has been the minimum required C standard version for Zephyr since
its inception. After multiple attempts and discussions, a decision has
been made to upgrade to C17 going forward.
This commits replaces the default C standard from C99 to C17 in the
configuration and build system, and deprecates support for the older
standards.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

pdgendt commented on doc/releases/migration-guide-4.4.rst:
```diff
@@ -23,6 +23,11 @@ the :ref:`release notes<zephyr_4.4>`.
+* Zephyr now officially enforces C17 (ISO/IEC 9899:2018) as its minimum required
+  C standard version.  If your toolchain does not support this standard you will
+  need to use one of the existing and now deprecated options:
+  :kconfig:option:`CONFIG_STD_C99` or :kconfig:option:`CONFIG_STD_C11`.
```
Wording does not match the contents of the PR. C17 is the default, rather than enforced.

nordicjm commented on Kconfig.zephyr:
```diff
@@ -381,27 +382,28 @@ config REQUIRES_STD_C23
 	  as ISO/IEC 9899:1990. This version is known as "ANSI C".
 
 config STD_C99
 	bool "C99"
```
as above

carlescufi commented:
Last push: Add `[DEPRECATED]` to the Kconfig description text.

pdgendt commented on Kconfig.zephyr:
```diff
@@ -381,27 +382,28 @@ config REQUIRES_STD_C23
 	  C Standards.
 
 config STD_C90
 	bool "C90"
```
Still missing here.
